cirq-core/cirq/protocols/commutes_protocol.py

```diff
@@ -78,8 +79,8 @@ def commutes(
     v2: Any,
     *,
     atol: Union[int, float] = 1e-8,
-    default: Union[bool, TDefault] = RaiseTypeErrorIfNotProvided,
-) -> Union[bool, TDefault]:
+    default: Union[bool, _TRaiseTypeErrorIfNotProvided, None] = RaiseTypeErrorIfNotProvided,
+) -> Optional[bool]:
```

Comment:
I think we use type overloads here, something like:
```python
RaiseTypeErrorIfNotProvided = object()

@overload
def commutes(v1: Any, v2: Any, *, atol: Union[int, float] = 1e-8) -> bool:
    ...

@overload
def commutes(v1: Any, v2: Any, *, atol: Union[int, float] = 1e-8, default: TDefault) -> Union[bool, TDefault]:
    ...

def commutes(v1: Any, v2: Any, *, atol: Union[int, float] = 1e-8, default: Any = RaiseTypeErrorIfNotProvided) -> Any:
    ...
    if default is not RaiseTypeErrorIfNotProvided:
        return default
    raise TypeError(...)
```
This seems to do what we want:
```
a, b = object(), object()
reveal_type(commutes(a, b))
reveal_type(commutes(a, b, atol=1e-4))
reveal_type(commutes(a, b, default=None))
reveal_type(commutes(a, b, atol=1e-4, default=None))
reveal_type(commutes(a, b, default='foo'))
reveal_type(commutes(a, b, atol=1e-4, default='foo'))

# output from check/mypy:
cirq-core/cirq/protocols/commutes_protocol.py:169: note: Revealed type is 'builtins.bool'
cirq-core/cirq/protocols/commutes_protocol.py:170: note: Revealed type is 'builtins.bool'
cirq-core/cirq/protocols/commutes_protocol.py:171: note: Revealed type is 'Union[builtins.bool, None]'
cirq-core/cirq/protocols/commutes_protocol.py:172: note: Revealed type is 'Union[builtins.bool, None]'
cirq-core/cirq/protocols/commutes_protocol.py:173: note: Revealed type is 'Union[builtins.bool, builtins.str*]'
cirq-core/cirq/protocols/commutes_protocol.py:174: note: Revealed type is 'Union[builtins.bool, builtins.str*]'
```